pkg/snapshotterutil/sociutil.go

```diff
@@ -0,0 +1,125 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package snapshotterutil
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/containerd/nerdctl/pkg/imgutil"
+	"github.com/sirupsen/logrus"
+)
+
+// CreateSoci creates a SOCI index(`rawRef`)
+func CreateSoci(rawRef string) error {
+	sociExecutable, err := exec.LookPath("soci")
```

Comment:
Is shelling out to the SOCI CLI a design goal or is it just because the CLI has non-trivial logic?

I would think using SOCI as a library here would be better long term.

Reply:
If I understand correctly, the pros of using SOCI library include:

- Does not require SOCI CLI to be installed in Nerdctl env
- Cleaner code to call library instead of calling commands

However, there is already a common practice of calling commands of dependencies in Nerdctl, such as Buildkit, Cosign, Notation, and `soci-snapshotter-grpc` is anyways needed to be pre-configured as part of SOCI provisioning. So the two pros don't seem significant.

Here are the reasons of leaning to using SOCI CLI instead of library for now.

- SOCI CLI has well-defined documentation and the backward compatibility is respected by default. The library code doesn't have documentation and it is not clear about how much backward compatibility it will have. It may also be difficult for SOCI maintainers to promise and define that.
- SOCI CLI layer has some business logic which can call multiple methods of library code, which is tedious to duplicate in Nerdctl. 

For both above, some refactoring in SOCI similar to [what Nerdctl is doing](https://github.com/containerd/nerdctl/issues/1680) could help (with that, the library code can include all the business logic and can be clearly defined and promised backward compatibility) but may not deserve.